dot/state/pruner/pruner.go

```diff
@@ -0,0 +1,354 @@
+	// Hash of keys that are deleted from state trie of the block
+	deletedKeys []common.Hash
+}
+
+type journalKey struct {
+	blockNum  int64
+	blockHash common.Hash
+}
+
+func newJournalRecord(hash common.Hash, insertedKeys, deletedKeys []common.Hash) *journalRecord {
+	return &journalRecord{
+		blockHash:    hash,
+		insertedKeys: insertedKeys,
+		deletedKeys:  deletedKeys,
+	}
+}
+
+// NewFullNode creates a Pruner for full node.
+func NewFullNode(db, storageDB chaindb.Database, retainBlocks int64, l log.Logger) (Pruner, error) {
+	p := &FullNode{
+		deathList:    make([]deathRow, 0),
+		deathIndex:   make(map[common.Hash]int64),
+		storageDB:    storageDB,
+		journalDB:    chaindb.NewTable(db, journalPrefix),
+		retainBlocks: retainBlocks,
+		logger:       l,
+	}
+
+	blockNum, err := p.getLastPrunedIndex()
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("last pruned block", "block num", blockNum)
+	blockNum++
+
+	p.pendingNumber = blockNum
+
+	err = p.loadDeathList()
+	if err != nil {
+		return nil, err
+	}
+
+	go p.start()
+
+	return p, nil
+}
+
+// StoreJournalRecord stores journal record into DB and add deathRow into deathList
+func (p *FullNode) StoreJournalRecord(deleted, inserted []common.Hash, blockHash common.Hash, blockNum int64) error {
+	jr := newJournalRecord(blockHash, inserted, deleted)
+
+	key := &journalKey{blockNum, blockHash}
+	err := p.storeJournal(key, jr)
+	if err != nil {
+		return fmt.Errorf("failed to store journal record for %d: %w", blockNum, err)
+	}
+
+	p.logger.Debug("journal record stored", "block num", blockNum)
+	p.addDeathRow(jr, blockNum)
+	return nil
+}
+
+func (p *FullNode) addDeathRow(jr *journalRecord, blockNum int64) {
+	if blockNum == 0 {
+		return
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	// The block is already pruned.
+	if blockNum < p.pendingNumber {
+		return
+	}
+
+	p.processInsertedKeys(jr.insertedKeys, jr.blockHash)
+
+	// add deleted keys from journal to death index
+	for _, k := range jr.deletedKeys {
+		p.deathIndex[k] = blockNum
+	}
+
+	deletedKeys := make(map[common.Hash]int64)
+	for _, data := range jr.deletedKeys {
+		deletedKeys[data] = blockNum
+	}
+
+	blockIndex := blockNum - p.pendingNumber
+	for idx := blockIndex - int64(len(p.deathList)); idx >= 0; idx-- {
+		p.deathList = append(p.deathList, deathRow{})
+	}
+
+	record := &deathRecord{
+		blockHash:   jr.blockHash,
+		deletedKeys: deletedKeys,
+	}
+
+	// add deathRow to deathList
+	p.deathList[blockIndex] = append(p.deathList[blockIndex], record)
+}
+
+// Remove re-inserted keys
+func (p *FullNode) processInsertedKeys(insKeys []common.Hash, blockHash common.Hash) {
+	for _, k := range insKeys {
+		num, ok := p.deathIndex[k]
+		if !ok {
+			continue
+		}
+		records := p.deathList[num-p.pendingNumber]
+		for _, v := range records {
+			if v.blockHash == blockHash {
+				delete(v.deletedKeys, k)
+			}
+		}
+		delete(p.deathIndex, k)
+	}
+}
+
+func (p *FullNode) start() {
+	p.logger.Info("pruning started")
```

Comment:
change to debug

Reply:
This log is printed only once during the entire runtime of the entire program and indicates that the pruning is started.

Reply:
same as above, I don't think this is necessary to log as info